flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -523,15 +557,18 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 					return;
 				}
 
-				if (expectedSequenceNumber != sequenceNumber) {
```

Comment:
`RemoteInputChannel#onBuffer` is only called by the same netty thread. Because `expectedSequenceNumber` is only updated and read from this method, it is thread safe. `onError` is also thread safe (setting atomic).
I extracted that portion exactly to limit the scope of synchronized.